abdiff/core/run_ab_transforms.py

```diff
@@ -1 +1,183 @@
+
+    # get run ID (used to label containers)
+    run_id = str(run_directory).split("/")[-1]
+
+    # create directories for 'transformed' outputs
+    transformed_directory_a, transformed_directory_b = create_directories(run_directory)
+
+    # create list of run configs
+    run_configs = [
+        (image_name_a, transformed_directory_a),
+        (image_name_b, transformed_directory_b),
+    ]
+
+    # transform input files using images A and B via Docker containers
+    # save transformed record JSON files in corresponding directory
+    containers = []
+    for input_file in input_files:
+        for docker_image, transformed_directory in run_configs:
+            container = run_docker_container(
+                docker_image, transformed_directory, run_id, input_file
+            )
+            containers.append(container)
+
+    # wait for all container runs to complete (status = "exited")
+    wait_for_containers(containers)
+    logger.info(f"All {len(containers)} containers have exited.")
+
+    # get logs from all Docker containers
+    log_file = f"{run_directory}/transformed/logs.txt"
+    with open(Path(log_file), "wb") as file:
+        for container in containers:
+            header = (
+                f"docker_image: {container.labels['docker_image']} | "
+                f"source: {container.labels['source']} | "
+                f"input_file: {container.labels['input_file']}\n"
+            )
+            file.write(header.encode())
+            file.writelines(container.logs(stream=True))
+            file.write(b"\n")
+    logger.info(f"Log file created: {log_file}")
+
+    # update run.json with list of input files
+    update_or_create_run_json(run_directory, new_data={"input_files": input_files})
+
+    elapsed_time = perf_counter() - start_time
+    logger.info(
+        "Total time to complete process: %s", str(timedelta(seconds=elapsed_time))
+    )
+    return get_transformed_files(run_directory)
+
+
+def run_docker_container(
+    docker_image: str,
+    transformed_directory: str,
+    run_id: str,
+    input_file: str,
+) -> Container:
+    """Run transmogrifier via Docker container to transform input file."""
+    source, output_filename = parse_transform_details_from_extract_filename(input_file)
+    container = DockerClient.containers.run(
+        docker_image,
+        command=[
+            f"--input-file={input_file}",
+            f"--output-file=/tmp/{output_filename}",
+            f"--source={source}",
+        ],
+        detach=True,
+        environment={
+            "AWS_ACCESS_KEY_ID": CONFIG.AWS_ACCESS_KEY_ID,
+            "AWS_SECRET_ACCESS_KEY": CONFIG.AWS_SECRET_ACCESS_KEY,
+            "AWS_SESSION_TOKEN": CONFIG.AWS_SESSION_TOKEN,
+        },
+        labels={
+            "docker_image": docker_image,
+            "source": source,
+            "input_file": input_file,
+            "run": run_id,
+        },
+        volumes=[f"{transformed_directory}:/tmp"],
```

Comment:
Happy to wait for the "real" PR to dig into this if you'd like, but in some local testing I _think_ this has to be an absolute path to mount into the container.  

We could note this somewhere and make it the responsibility of the calling context to _always_ pass an absolute directory for the `run_directory` to the parent `run_ab_transforms` function. In this way, we could be sure that the calculated transformed directory(s) are absolute.

Or -- don't know offhand -- maybe python has a way to take a relative directory, and get the absolute path from that?

Reply:
Now that I've read through this closer, I see how `run_ab_transforms()` will pass an absolute path to the `run_docker_container` helper function when it runs.

That said, I did learn this is a thing, which could be an optional safeguard to add to the `run_docker_container()` function to ensure any path that is mounted is absolute:

```python
import os
absolute_path = os.path.abspath(relative_path)
```

Reply:
Noting a quasi-related comment below: https://github.com/MITLibraries/transmogrifier-ab-diff/pull/6#discussion_r1755319275

Reply:
Just to confirm: 

* If given a path that begins with a forward slash, `os.path.abspath` interprets the input as an absolute path. 
* If given a path that **does not** begin with a forward slash, `os.path.abspath` interprets the input as a relative path and returns a "normalized" path that **joins the current working directory with the input path**. 

Therefore, if we use `os.path.abspath`, we need to make sure that our current working directory is set to the base directory of the `transmogrifier-ab-diff` whenever we run `run_ab_transforms`. Is this how you understand it as well? 

Just wanted to make sure but totally on board to use `os.path.abspath` as needed!

Reply:
> Therefore, if we use os.path.abspath, we need to make sure that our current working directory is set to the base directory of the transmogrifier-ab-diff whenever we run run_ab_transforms. Is this how you understand it as well?

Kind of?  I think that 99% of our CLI apps assume that you are invoking them from the root of the project.  All of our relative imports are setup this way.  So I think it's pretty much implied this is what would happen.  

Related, no matter where you are, if you pass a relative path for something it must be relative to _where you are_.  So even then, `os.path.abspath()` would still work, as it'd be the (assumed correct) relative path + wherever you are.

To be clear: the use of `os.path.relpath()` or `os.path.abspath()` should **not** require ever doing any kind of `sys.path` setting or changing the working directory.  The real benefit of using `os.path.abspath()` would be: **if a relative path is used when identifying the job or run directory, it would be converted into an absolute path before mounting that into the Docker containers**.

Happy to talk more though!